Rerank: numberOfResults greater than number of sources

### Package

langchain-aws

### Checked other resources

- [x] I added a descriptive title to this issue
- [x] I searched the LangChain documentation with the integrated search
- [x] I used the GitHub search to find a similar issue and didn't find it
- [x] I am sure this is a bug and not a question or request for help

### Example Code

```python
from langchain_aws import BedrockRerank
from langchain_classic.retrievers.contextual_compression import (
    ContextualCompressionRetriever,
)

# pseudocode
retriever = vector_store.as_retriever(
        search_kwargs={
            "filter": ...,
            "k": get_settings().RETRIEVER_NUM_RESULTS,
            "score_threshold": get_settings().RETRIEVER_SCORE_THRESHOLD,
        }
    )

compressor = BedrockRerank(model_arn=f"arn:aws:bedrock:{get_settings().AWS_BEDROCK_REGION}::foundation-model/cohere.rerank-v3-5:0", top_n=5)

reranked_retriever = ContextualCompressionRetriever(
        base_compressor=compressor, base_retriever=retriever
    )
```

### Error Message and Stack Trace (if applicable)

```shell
2025-11-13 13:38:43,241 [ERROR] app.agents.tools.ubi_rag_tool:256: Error in event_rag_tool: An error occurred (ValidationException) when calling the Rerank operation: Cannot provide numberOfResults value that exceeds the number of sources. traceback: Traceback (most recent call last):
  File "/app/app/agents/tools/ubi_rag_tool.py", line 241, in search_event_tool
    docs = retriever.invoke(query, config)
  File "/app/.venv/lib/python3.13/site-packages/langchain_core/retrievers.py", line 216, in invoke
    result = self._get_relevant_documents(
        input, run_manager=run_manager, **kwargs_
    )
  File "/app/.venv/lib/python3.13/site-packages/langchain_classic/retrievers/contextual_compression.py", line 40, in _get_relevant_documents
    compressed_docs = self.base_compressor.compress_documents(
        docs,
        query,
        callbacks=run_manager.get_child(),
    )
  File "/app/.venv/lib/python3.13/site-packages/langchain_aws/document_compressors/rerank.py", line 187, in compress_documents
    for res in self.rerank(documents, query):
               ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.13/site-packages/langchain_aws/document_compressors/rerank.py", line 158, in rerank
    response = self.client.rerank(**request_body)
  File "/app/.venv/lib/python3.13/site-packages/botocore/client.py", line 602, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.13/site-packages/botocore/context.py", line 123, in wrapper
    return func(*args, **kwargs)
  File "/app/.venv/lib/python3.13/site-packages/botocore/client.py", line 1078, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.ValidationException: An error occurred (ValidationException) when calling the Rerank operation: Cannot provide numberOfResults value that exceeds the number of sources.
```

### Description

in `rerank.py` it is not checked if top_n is less than `len(serialized_documents)`.
Bedrock does not accept this. and while numberOfResults is in itself an "optional" parameter: https://docs.aws.amazon.com/bedrock/latest/APIReference/API_agent-runtime_BedrockRerankingConfiguration.html 

i cannot overwrite it so I do the top_n results array splice myself. after the rerank operation.

this is the call to bedrock
```
request_body = {
            "queries": [{"textQuery": {"text": query}, "type": "TEXT"}],
            "rerankingConfiguration": {
                "bedrockRerankingConfiguration": {
                    "modelConfiguration": {
                        "modelArn": self.model_arn,
                        "additionalModelRequestFields": additional_model_request_fields
                        or {},
                    },
                    "numberOfResults": top_n or self.top_n,
                },
                "type": "BEDROCK_RERANKING_MODEL",
            },
            "sources": [
                {"inlineDocumentSource": doc, "type": "INLINE"}
                for doc in serialized_documents
            ],
        }
```

whereas
```
top_n: Optional[int] = 3
    """Number of documents to return."""
```

my proposal:
```
num_results = min(top_n or self.top_n, len(serialized_documents))
```

I previously upgraded to langchain v1 and to bedrock rerank. Before that I never encoutered this bug with another reranker (used cohereRerank previously) so I did not expect this.

from my perspective changing this would require implementing the BedrockRerank myself.

--- a/libs/aws/langchain_aws/document_compressors/rerank.py
+++ b/libs/aws/langchain_aws/document_compressors/rerank.py
@@ -126,6 +126,10 @@ class BedrockRerank(BaseDocumentCompressor):
         if len(documents) == 0:
             return []
 
+        effective_top_n = top_n if top_n is not None else self.top_n
+        if effective_top_n is not None:
+            effective_top_n = min(effective_top_n, len(documents))
+
         # Serialize documents for the Bedrock API
         serialized_documents = [
             {"textDocument": {"text": doc.page_content}, "type": "TEXT"}
@@ -145,7 +149,7 @@ class BedrockRerank(BaseDocumentCompressor):
                         "additionalModelRequestFields": additional_model_request_fields
                         or {},
                     },
-                    "numberOfResults": top_n or self.top_n,
+                    "numberOfResults": effective_top_n,
                 },
                 "type": "BEDROCK_RERANKING_MODEL",
             },

--- a/libs/aws/tests/unit_tests/document_compressors/test_rerank.py
+++ b/libs/aws/tests/unit_tests/document_compressors/test_rerank.py
@@ -69,3 +69,89 @@ def test_compress_documents(mock_rerank: MagicMock, reranker: BedrockRerank) -> 
     assert len(compressed_docs) == 2
     assert compressed_docs[0].metadata["relevance_score"] == 0.95
     assert compressed_docs[1].metadata["relevance_score"] == 0.85
+
+
+@patch("langchain_aws.utils.create_aws_client")
+def test_rerank_clamps_top_n_to_document_count(
+    mock_create_client: MagicMock,
+) -> None:
+    mock_client = MagicMock()
+    mock_create_client.return_value = mock_client
+
+    captured_request: dict = {}
+
+    def mock_rerank(**kwargs: dict) -> dict:
+        captured_request.update(kwargs)
+        num_results = kwargs["rerankingConfiguration"]["bedrockRerankingConfiguration"][
+            "numberOfResults"
+        ]
+        return {
+            "results": [
+                {"index": i, "relevanceScore": 0.9 - (i * 0.1)}
+                for i in range(num_results)
+            ]
+        }
+
+    mock_client.rerank = mock_rerank
+
+    reranker = BedrockRerank(
+        model_arn="arn:aws:bedrock:us-west-2::foundation-model/amazon.rerank-v1:0",
+        region_name="us-east-1",
+        top_n=10,
+        client=mock_client,
+    )
+
+    documents = [
+        Document(page_content="Doc 1"),
+        Document(page_content="Doc 2"),
+        Document(page_content="Doc 3"),
+    ]
+
+    results = reranker.rerank(documents, query="test query")
+
+    actual_num_results = captured_request["rerankingConfiguration"][
+        "bedrockRerankingConfiguration"
+    ]["numberOfResults"]
+    assert actual_num_results == 3
+    assert len(results) == 3
+
+
+@patch("langchain_aws.utils.create_aws_client")
+def test_rerank_top_n_override_also_clamped(
+    mock_create_client: MagicMock,
+) -> None:
+    mock_client = MagicMock()
+    mock_create_client.return_value = mock_client
+
+    captured_request: dict = {}
+
+    def mock_rerank(**kwargs: dict) -> dict:
+        captured_request.update(kwargs)
+        num_results = kwargs["rerankingConfiguration"]["bedrockRerankingConfiguration"][
+            "numberOfResults"
+        ]
+        return {
+            "results": [
+                {"index": i, "relevanceScore": 0.9 - (i * 0.1)}
+                for i in range(num_results)
+            ]
+        }
+
+    mock_client.rerank = mock_rerank
+
+    reranker = BedrockRerank(
+        model_arn="arn:aws:bedrock:us-west-2::foundation-model/amazon.rerank-v1:0",
+        region_name="us-east-1",
+        top_n=2,
+        client=mock_client,
+    )
+
+    documents = [Document(page_content="Doc 1"), Document(page_content="Doc 2")]
+
+    results = reranker.rerank(documents, query="test query", top_n=100)
+
+    actual_num_results = captured_request["rerankingConfiguration"][
+        "bedrockRerankingConfiguration"
+    ]["numberOfResults"]
+    assert actual_num_results == 2
+    assert len(results) == 2

